packages/server/src/unstable-core-do-not-import/stream/sse.ts

```diff
@@ -59,108 +63,71 @@ type SSEvent = Partial<{
-    comment: 'connected',
-  });
+  stream.controller.enqueue({ comment: 'connected' });
 
-  const { serialize = (v) => v } = opts;
+  const { serialize = identity } = opts;
 
   const ping: Required<PingOptions> = {
     enabled: opts.ping?.enabled ?? false,
     intervalMs: opts.ping?.intervalMs ?? 1000,
   };
 
   run(async () => {
-    const iterator = opts.data[Symbol.asyncIterator]();
-
-    const closedPromise = stream.cancelledPromise.then(() => 'closed' as const);
-    const maxDurationPromise = createTimeoutPromise(
-      opts.maxDurationMs ?? Infinity,
-      'maxDuration' as const,
-    );
-
-    let nextPromise = iterator.next();
-
-    while (true) {
-      const pingPromise = createTimeoutPromise(
-        ping.enabled ? ping.intervalMs : Infinity,
-        'ping' as const,
-      );
-      const next = await Promise.race([
-        nextPromise.catch((err) => {
-          if (err instanceof Error && err.name === 'AbortError') {
-            // Ensures that aborting the request doesn't throw an error
-            return 'aborted' as const;
-          }
-          return getTRPCErrorFromUnknown(err);
-        }),
-        pingPromise.promise,
-        closedPromise,
-        maxDurationPromise.promise,
-      ]);
-
-      pingPromise.clear();
-      if (next === 'aborted') {
-        break;
-      }
-      if (next === 'closed') {
-        break;
-      }
-      if (next === 'maxDuration') {
-        break;
-      }
+    let iterable = withCancel(opts.data, stream.cancelledPromise);
 
-      if (next === 'ping') {
-        stream.controller.enqueue({
-          comment: 'ping',
-        });
-        continue;
-      }
+    if (opts.emitAndEndImmediately) {
+      iterable = takeWithGrace(iterable, { count: 1, gracePeriodMs: 1 });
+    }
 
-      if (next instanceof Error) {
-        const data = opts.formatError
-          ? opts.formatError({ error: next })
-          : null;
-        stream.controller.enqueue({
-          event: SERIALIZED_ERROR_EVENT,
-          data: JSON.stringify(serialize(data)),
-        });
-        break;
-      }
-      if (next.done) {
-        break;
-      }
+    let maxDurationTimer: PromiseTimer | null = null;
+    if (
+      opts.maxDurationMs != null &&
+      opts.maxDurationMs > 0 &&
+      opts.maxDurationMs !== Infinity
+    ) {
+      maxDurationTimer = createPromiseTimer(opts.maxDurationMs).start();
+      iterable = withCancel(iterable, maxDurationTimer.promise);
+    }
 
-      const value = next.value;
+    if (ping.enabled && ping.intervalMs !== Infinity && ping.intervalMs > 0) {
+      iterable = withPing(iterable, ping.intervalMs);
+    }
 
-      const chunk: SSEvent = isTrackedEnvelope(value)
-        ? {
-            id: value[0],
-            data: value[1],
-          }
-        : {
-            data: value,
-          };
-      if ('data' in chunk) {
-        chunk.data = JSON.stringify(serialize(chunk.data));
-      }
+    try {
+      for await (const value of iterable) {
```

Comment:
I've just added type-safety to avoid that mix of `unknown | typeof PING_SYM ⇒ unknown`. So, now it's exhaustive by typescript nature.

I don't like too much of a `switch` construct and don't see to much more than a PING to be added to iterables. But of course, you could now do `switch (chunk.value) { case PING_SYM: ...; default: ...; }` if you like that construct.